src/bit-systems/loop-animation.ts

```diff
@@ -0,0 +1,87 @@
+import { addComponent, defineQuery, enterQuery, exitQuery, hasComponent } from "bitecs";
+import { AnimationAction, AnimationClip, AnimationMixer, LoopRepeat } from "three";
+import { MixerAnimatable, MixerAnimatableData, LoopAnimation, LoopAnimationData, Object3DTag } from "../bit-components";
+import { HubsWorld } from "../app";
+
+const loopAnimationQuery = defineQuery([MixerAnimatable, LoopAnimation, Object3DTag]);
```

Comment:
This works good for JSX entities like the LoadingObject where the `MixerAnimatable` and `LoopAnimation` components are both added to the root object where the animations are but when a glTF model is loaded the `MixerAnimatableInitialize` (and hence the `MixerAnimatable` component) is added to the glTF root node but the `LoopAnimation` component can be in a child node so `loopAnimationQuery` doesn't iterate over the right `LoopAnimation` component.

My understanding is that we need to add a `MixerAnimatable` and create a mixer for any object with animations and every child with a `LoopAnimation` component will look for a parent with a `MixerAnimatable` and get the mixer and animations and then create `AnimationActions` based on its `LoopAnimation`  parameters.

My other comment referred to this, when a model is loaded in `inflateModel` if there are any animations we can already create a mixer and link the animations, something like:

```typescript
if (model.animations !== undefined && model.animations.length > 0) {
    const mixer = new AnimationMixer(model);
    MixerAnimatableData.set(rootEid, {
      mixer,
      animations: model.animations
    });
    addComponent(world, MixerAnimatable, rootEid);
    if (!findChildWithComponent(world, LoopAnimation, rootEid)) {
      inflateLoopAnimation(world, rootEid, {});
    }
  }
```

and then during the `LoopAnimation` query enter loop retrieve the mixer and animations from closest parent:
```typescript
const mixerEid = findAncestorWithComponent(world, MixerAnimatable, eid);
const { mixer, animations } = MixerAnimatableData.get(mixerEid)!;
```

Reply:
I checked through the old code and you seem to be right. It seems that `AnimationMixer` needs to be found from ancestors and root entity needs to be aware of `LoopAnimation`s in its children. I don't really like this design, but we need to make the new code compatible with the old code for now. I will update the code.

Reply:
We have talked about this privately but for future reference about how the `loop-animation` component works.

> 1. Is loop-animation Hubs component a glTF node level extension, not a glTF scene level extension?

As of this writing the `loop-animation` component can the attached to any mesh node. In Blender we expose the animation actions/nla tracks of that mesh for selection so each mesh with a `loop-animation` will only reference it's own animations and it can't reference animations in other meshes. Same applies for Spoke, only meshes with animations will show the Loop Animation component options and list that mesh animations.

> 2. Can there be multiple loop-animations in glTF nodes in a single glTF, for example gltf.nodes[0], gltf.nodes[1], and gltf.nodes[2] have loop-animation component?

Yes, any mesh node can have one.

> 3. If 2. is yes, can loop-animation components have different component parameters from each other?

Yes, each `loop-animation` component can have their own independent parameters.

> 4. If 3. is yes, can loop-animation component parameters that are contradict from each other be valid, for example gltf.nodes[0].loop-animation = {activeClipIndex: 0, paused: false}, gltf.nodes[1].loop-animation = {activeClipIndex: 0, paused: true}?

No, each `loop-animation` component will only reference it's own meshes animations and each mesh node can on ly have one `loop-animation` component so conflict is not possible.

Reply:
Thanks for the explanation. My impression is Hubs animation system assumes a little bit too much implied rules in glTF, for example it assumes that an animation clip consists of a ceratain root glTF node and its child nodes but there is no such a limitation in the glTF animation specification.

I think it would be better to let `loop-animation` component be aligned with glTF spec more and remove such implied rules. Probably the component specification will be more straightforward and less problematic, for example when processing a Hubs glTF assets with external glTF tools.

If we want to keep the animation representation in Blender, for example binding a `loop-animation` copmponent to a certain node, we may add info needed in glTF extra field.

It isn't the best time to revisit the `loop-animation` component spec now. We may revisit the spec later.